### ISSUE TITLE
dimensionality of class KPconv

### ISSUE THREAD

**Issue by pixar0407** [HuguesTHOMAS/KPConv-PyTorch#92]
Hello, I have confusion about the dimension of KPconv.
I am trying to match the formulation on the paper and the code.

 
at models/blocks.py
we have def forward at 237 line.

in here, we get
* all_weights : [n_points, n_kpoints, n_neighbors]
* neighb_x : [n_points, n_neighbors, in_fdim]

> weighted_features = torch.matmul(all_weights, neighb_x)

* weighted_features : [n_points, n_kpoints, in_fdim]

> weighted_features = weighted_features.permute((1, 0, 2))

* weighted_features : [n_kpoints,n_points, in_fdim]
* self.weights : [n_kpoints,in_fdim,out_fim]

> kernel_outputs = torch.matmul(weighted_features, self.weights)

* kernel_outputs : [n_kpoints,n_points,out_fim]

> torch.sum(kernel_outputs, dim=0)

final output : [n_points,out_fim]

I can understand the calculation in the code. But it seems little different with the equation (1), (2), (3) - 'Kernel Point Convolution' in the paper section 3.1.

It seems the matrix computation sequence has been changed.
So, does the code represent the equation (1), (2), (3) correctly or do i mis-understand?

thank you.

**Comment by HuguesTHOMAS** [comment]
Hi @pixar0407,

Thanks for your interest in my work. You are right the code does not follow the same logic as the paper and that can be surprising. But, actually, if you do the math, the sequence of operations in the code gives the exact same result as what is described in the paper. It basically comes down to an inversion of the summation indexes. Here is the proof:

In the paper if you combine (1) and (2) you get: `sum_i[ sum_k( h_ik * W_k * f_i ) ]`

The two sums can be inverted and then the result is equivalent to: `sum_k[ sum_i( h_ik * f_i ) * W_k ]` which is what is implemented in the code as it is the most efficient.

In conclusion, you can either consider the convolution to be the scaled kernel weight multiplying the features (as in the paper) or the scaled features multiplied by the kernel weights (as in the code).

I hope this clarifies things for you.

Best,
Hugues

**Comment by pixar0407** [comment]
Hello, @HuguesTHOMAS.

Oh! your explanation seems quite clear!
The concept "the scaled features multiplied by the kernel weights" is amazing.

Anyway, thank you for the reply and the quality code!
have a nice day.
